airflow/models/trigger.py

```diff
@@ -0,0 +1,161 @@
+    kwargs = Column(ExtendedJSON, nullable=False)
+    created_date = Column(UtcDateTime, nullable=False)
+
+    def __init__(
+        self, classpath: str, kwargs: Dict[str, Any], created_date: Optional[datetime.datetime] = None
+    ):
+        super().__init__()
+        self.classpath = classpath
+        self.kwargs = kwargs
+        self.created_date = created_date or timezone.utcnow()
+
+    @classmethod
+    @provide_session
+    def runnable_ids(
+        cls, session=None, partition_ids: Optional[List[int]] = None, partition_total: Optional[int] = None
+    ):  # pylint: disable=unused-argument
+        """
+        Returns all "runnable" triggers IDs, optionally filtering down by partition.
+
+        This is a pretty basic partition algorithm for now, but it does the job.
+        """
+        # NOTE: It's possible in future that we could try and pre-calculate a
+        # partition entry in a large virtual ring (e.g. 4096 buckets) and store
+        # that in the DB for more direct querying, but for now Jump is fast
+        # enough of a hash to do this all locally - about 0.1s per million hashes
+
+        # Retrieve all IDs first
+        trigger_ids = [row[0] for row in session.query(cls.id).all()]
+
+        # Short-circuit for "no partitioning"
+        if partition_ids is None or partition_total is None:
+            return trigger_ids
+
+        # Go through and map each trigger ID to a partition number,
+        # using a quick, consistent hash (Jump), keeping only the ones that
+        # match one of our partition IDs
+        return [x for x in trigger_ids if jump.hash(x, partition_total) + 1 in partition_ids]
+
+    @classmethod
+    def from_object(cls, trigger: BaseTrigger):
+        """
+        Alternative constructor that creates a trigger row based directly
+        off of a Trigger object.
+        """
+        classpath, kwargs = trigger.serialize()
+        return cls(classpath=classpath, kwargs=kwargs)
+
+    @classmethod
+    @provide_session
+    def bulk_fetch(cls, ids: List[int], session=None) -> Dict[int, "Trigger"]:
+        """
+        Fetches all of the Triggers by ID and returns a dict mapping
+        ID -> Trigger instance
+        """
+        return {obj.id: obj for obj in session.query(cls).filter(cls.id.in_(ids)).all()}
+
+    @classmethod
+    @provide_session
+    def clean_unused(cls, session=None):
+        """
+        Deletes all triggers that have no tasks/DAGs dependent on them
+        (triggers have a one-to-many relationship to both)
+        """
+        # Update all task instances with trigger IDs that are not DEFERRED to remove them
+        session.query(TaskInstance).filter(
+            TaskInstance.state != State.DEFERRED, TaskInstance.trigger_id.isnot(None)
+        ).update({TaskInstance.trigger_id: None})
+        # Get all triggers that have no task instances depending on them...
+        ids = [
+            x[0]
+            for x in (
```

Comment:
```suggestion
            trigger_id
            for (trigger_id,) in (
```

I think is slightly easier to read.

Reply:
You're not wrong.